utils/loggers/mlflow/mlflow_utils.py

```diff
@@ -0,0 +1,134 @@
+            LOGGER.warning(f"{prefix}Continuining without Mlflow")
+            self.mlflow = None
+            self.mlflow_active_run = None
+
+    def setup(self, opt: Namespace) -> None:
+        if opt.weights is not None and str(opt.weights).strip() != "":
+            model_name = Path(opt.weights).stem
+        else:
+            model_name = "yolov5"
+        self.model_name = model_name
+        self.weights = Path(opt.weights)
+        self.client = mlflow.tracking.MlflowClient()
+        self.log_params(vars(opt))
+        self.log_metrics(vars(opt), is_param=True)
+
+    @staticmethod
+    def _flatten_params(params_dict, parent_key="", sep="/"):
+        """Static method to flatten configs for logging to mlflow"""
+        items = []
+        for key, value in params_dict.items():
+            new_key = parent_key + sep + key if parent_key else key
+            if isinstance(value, MutableMapping):
+                items.extend(MlflowLogger._flatten_params(value, new_key, sep).items())
+            else:
+                items.append((new_key, value))
+        return dict(items)
+
+    def log_artifacts(self, artifact: Path, relpath: str = None) -> None:
+        """Member function to log artifacts (either directory or single item).
+
+        Args:
+            artifact (Path): File or folder to be logged
+            relpath (str): Name (or path) relative to experiment for logging artifact in mlflow
+        """
+        if not isinstance(artifact, Path):
+            artifact = Path(artifact)
+        if artifact.is_dir():
+            self.mlflow.log_artifacts(f"{artifact.resolve()}/", artifact_path=str(artifact.stem))
+        else:
+            self.mlflow.log_artifact(str(artifact.resolve()), artifact_path=relpath)
+
+    def log_model(self, model_path: Path, model_name: str = None) -> None:
+        """Member function to log model as an Mlflow model.
+
+        Args:
+            model_path: Path to the model .pt being logged
+            model_name: Name (or path) relative to experiment for logging model in mlflow
+        """
+        self.mlflow.pyfunc.log_model(artifact_path=self.model_name if model_name is None else model_name,
+                                     code_path=[str(ROOT.resolve())],
+                                     artifacts={"model_path": str(model_path.resolve())},
+                                     python_model=self.mlflow.pyfunc.PythonModel())
+
+    def log_params(self, params: Dict[str, Any]) -> None:
+        """Member funtion to log parameters.
+        Mlflow doesn't have mutable parameters and so this function is used
+        only to log initial training parameters.
+
+        Args:
+            params (Dict[str, Any]): Parameters as dict
+        """
+        try:
+            flattened_params = MlflowLogger._flatten_params(params_dict=params)
+            run = self.client.get_run(run_id=self.run_id)
+            logged_params = run.data.params
+            [
```

Comment:
May 26th, 2023, the first iteration in the list comprehension fails, triggering the generic try and except, and skipping every param, perhaps change the list comprehension to a for loop, have the try and except inside the for loop?

```python
        newly_logged = []
        for key, value in flattened_params.items():
            if (
                key in logged_params or not value
            ):  # TODO find out if mlflow.log_param 'Appends' or 'Overwrites'
                continue
            try:
                str(value).strip()
                assert value
                newly_logged.append(mlflow.log_param(key, str(value)))
            except AssertionError:
                LOGGER.info("%s, not logging None value: %s", LOG_PREFIX, key)